components/sup/src/manager/service/supervisor.rs

```diff
@@ -35,26 +36,53 @@ static LOGKEY: &str = "SV";
 
 #[derive(Debug)]
 pub struct Supervisor {
-    pub preamble:      String,
-    pub state:         ProcessState,
+    service_group: ServiceGroup,
+    state:         ProcessState,
+    // TODO (CM): make this private
     pub state_entered: Timespec,
-    pid:               Option<Pid>,
-    pid_file:          PathBuf,
+    pid: Option<Pid>,
+    /// If the Supervisor is being run with the PIDS_FROM_LAUNCHER
+    /// feature enabled, this will be `None`, otherwise it will be
+    /// `Some(path)`. Client code should use the `Some`/`None` status
+    /// of this field as an indicator of which mode the Supervisor is
+    /// running in.
+    pid_file: Option<PathBuf>,
 }
 
 impl Supervisor {
-    pub fn new(service_group: &ServiceGroup) -> Supervisor {
-        Supervisor { preamble:      service_group.to_string(),
-                     state:         ProcessState::Down,
+    /// Create a new instance for `service_group`.
+    ///
+    /// The `pid_source` governs how we determine the PID of the
+    /// supervised process. Once the PIDS_FROM_LAUNCHER feature flag
+    /// goes away, this can be removed.
+    pub fn new(service_group: &ServiceGroup, pid_source: ServicePidSource) -> Supervisor {
+        let pid_file = match pid_source {
+            ServicePidSource::Files => Some(fs::svc_pid_file(service_group.service())),
+            ServicePidSource::Launcher => None,
+        };
```

Comment:
I wouldn't necessarily change this since we think that the feature is going to be removed eventually.  But, if we think this is going to live in the code base a while, I'm wondering it it would make sense to move the behavior differences between the pid sources out of this code and out into a trait.

Reply:
If it was intended as a long-standing variation, then yeah, adding a bit more structure around this would be good.

I think that since this is more of a transition state, though, the amount of structure we've got now is probably good enough.

We _should_ draw up a concrete timeline for when that switchover should take place, though.